### PR TITLE
Unpin astropy-header and delay astropy import

### LINKED ISSUE #12646: MNT: Unpin pytest-astropy-header when 0.2.1 or later is released
So we can undo #12643 and part of #12639.

Also see astropy/pytest-astropy-header#41

### PR DESCRIPTION
**Pull request by saimn**
Alternative to  #12659

### Description

This pull request is to address ...

Fixes #12646

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -24,6 +22,8 @@ def pytest_configure(config):
     PYTEST_HEADER_MODULES['Scikit-image'] = 'skimage'
     PYTEST_HEADER_MODULES['asdf'] = 'asdf'
     PYTEST_HEADER_MODULES['pyarrow'] = 'pyarrow'
+
+    from astropy import __version__
```
Comment (pllim) [inline conftest.py:26]: I thought I tried this in an earlier attempt and I didn't work. But if it works now, then this is definitely cleaner.

**Comment by pllim** [comment]
Now it passes! 🤯

**Comment by pllim** [comment]
Hmm... Looks like the jobs still picking up `pytest-astropy-header==0.1.2`.

**Comment by pllim** [comment]
I rebased for you and now it is showing that the error is still there even with this patch.

**Comment by pllim** [comment]
@nstarman , do we really need a fake package in the test suite?

```
________________________ ERROR collecting test session _________________________
../../.tox/py310-test-cov/lib/python3.10/site-packages/astropy/cosmology/tests/mypackage/io/tests/conftest.py:7: in <module>
    from mypackage.io import ASTROPY_GE_5
E   ModuleNotFoundError: No module named 'mypackage'
```

**Comment by nstarman** [comment]
Hmm. Yes? If possible. 🤞 
`mypackage` is meant to demonstrate how to do Cosmology I/O. Portions of the code files are shown in the docs using ``literalinclude``. Including `mypackage` in ``cosmology/tests`` ensures that it is tested, which was a problem I encountered when `mypackage` was previously in ``docs/``.

**Comment by nstarman** [comment]
@pllim, with the new system, is it still possible? The `says.path` hack isn't working.

**Comment by pllim** [comment]
@saimn , what do you think?

@astrofrog , maybe we can move `mypackage` stuff out of core into your "integration testing" project?

**Comment by pllim** [comment]
As per dev telecon today (2022-01-11), here is the path forward that would unblock this PR:

1. @nstarman will work with @hamogu to create a new package under `astropy` org to represent `cosmology/tests/mypackage` outside of core lib.
2. Then `cosmology/tests/mypackage` will be removed from core lib.
3. Then this PR can rebase and does not have to worry about `conftest.py` in ``cosmology/tests/mypackage`.
4. Maybe: This new package might be a good candidate for https://github.com/astropy/integration-testing when @astrofrog can get around to it.

**Comment by saimn** [comment]
Hmm, actually `--import-mode=importlib` doesn't work. It also changes the import mode for test files (not only `conftest`) which makes it impossible to do relative imports from test files.

**Comment by saimn** [comment]
For reference, more details on the importlib option: https://docs.pytest.org/en/6.2.x/pythonpath.html 

> importlib: new in pytest-6.0, this mode uses importlib to import test modules. This gives full control over the import process, and doesn’t require changing sys.path or sys.modules at all.

> For this reason this doesn’t require test module names to be unique at all, but also **makes test modules non-importable by each other**. This was made possible in previous modes, for tests not residing in Python packages, because of the side-effects of changing sys.path and sys.modules mentioned above. Users which require this should turn their tests into proper packages instead.

> **We intend to make importlib the default in future releases**.

**Comment by pllim** [comment]
Do we really need relative imports anyway?

**Comment by saimn** [comment]
Maybe not but there are a lot of errors with the relative imports we have currently, and also a few other errors which I did not investigate.

**Comment by nstarman** [comment]
This makes moving `mypackage` a little less urgent, but I still like the ideas people came up with. A separate package or as part of `astropy-data` makes more sense than the current location.

**Comment by pllim** [comment]
I remember the good old times when `pytest` was just one file and didn't give us much trouble...

**Comment by saimn** [comment]
> I remember the good old times when pytest was just one file and didn't give us much trouble...

Well, I guess the import magic has been there since a long time... the real problem with tox/pytest is that it's painful to setup a workflow where you can test the installed code. The simplest way to achieve that in to use a "src layout" but people seem reluctant to that.

**Comment by pllim** [comment]
With all the open PRs, moving stuff to `src` is going to be painful...

**Pull request closed without merging**